src/node.rs

```diff
@@ -0,0 +1,169 @@
+use crate::config::ServerConfig;
+use crate::error::{ErrorKind, Result};
+use crate::secp::Commitment;
+
+use grin_api::client;
+use grin_api::json_rpc::{build_request, Request, Response};
+use grin_api::{BlockPrintable, LocatedTxKernel, OutputPrintable, OutputType, Tip};
+use grin_core::consensus::COINBASE_MATURITY;
+use grin_core::core::{Input, OutputFeatures, Transaction};
+use grin_util::ToHex;
+
+use serde_json::json;
+use std::net::SocketAddr;
+
+const ENDPOINT: &str = "/v2/foreign";
+
+#[derive(Clone)]
+pub struct HTTPNodeClient {
+    node_url: SocketAddr,
+    node_api_secret: Option<String>,
+}
+
+impl HTTPNodeClient {
+    /// Create a new client that will communicate with the given grin node
+    pub fn new(node_url: &SocketAddr, node_api_secret: Option<String>) -> HTTPNodeClient {
+        HTTPNodeClient {
+            node_url: node_url.to_owned(),
+            node_api_secret: node_api_secret,
+        }
+    }
+
+    fn send_json_request<D: serde::de::DeserializeOwned>(
+        &self,
+        method: &str,
+        params: &serde_json::Value,
+    ) -> Result<D> {
+        let url = format!("http://{}{}", self.node_url, ENDPOINT);
+        let req = build_request(method, params);
+        let res =
+            client::post::<Request, Response>(url.as_str(), self.node_api_secret.clone(), &req)?;
+        let parsed = res.clone().into_result()?;
+        Ok(parsed)
+    }
+}
+
+#[derive(Clone)]
+pub struct GrinNode {
+    client: HTTPNodeClient,
+}
+
+impl GrinNode {
+    pub fn new(server_config: &ServerConfig) -> GrinNode {
+        GrinNode { 
+            client: HTTPNodeClient::new(&server_config.grin_node_url, None),
+        }
+    }
+
+    // Checks whether a commitment is spendable at the block height provided
+    pub fn is_spendable(&self, output_commit: &Commitment, next_block_height: u64) -> Result<bool> {
+        let output = self.get_output(&output_commit)?;
+        if let Some(out) = output {
+            let is_coinbase = match out.output_type {
+                OutputType::Coinbase => true,
+                OutputType::Transaction => false,
+            };
+
+            if is_coinbase {
+                if let Some(block_height) = out.block_height {

```

Comment:
I think we could just forbid mixing of coinbases to avoid maturity checks and because self-spending them serves no purpose.

Reply:
I kept coinbase support since the work was already done, and it shouldn't negatively affect the swap in any way. But if there's a reason we really don't want to include coinbase outputs, it's easy enough to disallow them.

Reply:
I can see no reason to ever want to swap a coinbase, so I'd remove support if it simplifies the code.